### PR TITLE
subpackage index exports & migration to ts

### PR DESCRIPTION
**Pull request by mveres**

### CODE REVIEW

File: clipboard/dockit-clipboard.define.js [hunk 1]
```diff
@@ -1,3 +1,3 @@
-import { Clipboard } from './src/Clipboard.js';
+import { Clipboard } from './src/Clipboard';
```
Comment (bashmish) [inline clipboard/dockit-clipboard.define.js:1]: this is gonna break Custom Element Manifest generator, it requires extension, so please add an extension everywhere where you removed it

File: tsconfig.json [hunk 2]
```diff
@@ -9,6 +9,7 @@
     "module": "ESNext",
     "moduleResolution": "Node",
     "useDefineForClassFields": false,
+    "allowSyntheticDefaultImports": true,
```
Comment (bashmish) [inline tsconfig.json:12]: why is it needed?

File: clipboard/dockit-clipboard.define.js (hunk 1 above)
Comment (mveres) [reply clipboard/dockit-clipboard.define.js:1]: Is the extension mandatory in `index` files also?

File: tsconfig.json (hunk 2 above)
Comment (mveres) [reply tsconfig.json:12]:
```
Module 'dockit-core/node_modules/copy-to-clipboard/index' can only be default-imported using the 'allowSyntheticDefaultImports' flag
```
I haven't found other workarounds.

File: box/dockit-box.define.js [hunk 3]
```diff
@@ -1,2 +1,2 @@
-import { Box } from './src/Box.js';
+import { Box } from './src/Box.ts';
```
Comment (gqio) [inline box/dockit-box.define.js:1]: please do not use .ts extension it's not valid extension in import AFAICT.

File: clipboard/dockit-clipboard.define.js [hunk 4]
```diff
@@ -1,3 +1,3 @@
-import { Clipboard } from './src/Clipboard.js';
+import { Clipboard } from './src/Clipboard.ts';
```
Comment (gqio) [inline clipboard/dockit-clipboard.define.js:1]: same no need for ts extension here

File: note/dockit-note.define.js [hunk 5]
```diff
@@ -1,2 +1,2 @@
-import { Note } from './src';
+import { Note } from './src/Note.ts';
```
Comment (gqio) [inline note/dockit-note.define.js:1]: ts extension not needed

File: showcases/dockit-showcases.define.js [hunk 6]
```diff
@@ -1,3 +1,3 @@
-import { Showcases } from './src/Showcases.js';
+import { Showcases } from './src/Showcases.ts';
```
Comment (gqio) [inline showcases/dockit-showcases.define.js:1]: same

File: showcases/src/Showcases.ts [hunk 7]
```diff
@@ -39,7 +38,8 @@ export class Showcases extends HTMLElement {
       ${acc}
       ${renderComponent({
         componentClass,
-        [showcaseAttr]: showcase,
+        showcaseClass: !!showcaseClasses && showcase,
```
Comment (gqio) [inline showcases/src/Showcases.ts:41]: Sorry I don't understand this

**Review by gqio — COMMENTED** [review]
I think you can remove all ts extension when imported

File: showcases/src/Showcases.ts (hunk 7 above)
Comment (mveres) [reply showcases/src/Showcases.ts]: fixed, rephrased

File: clipboard/dockit-clipboard.define.js (hunk 1 above)
Comment (bashmish) [reply clipboard/dockit-clipboard.define.js]: it's mandatory to track how CE classes are constructed and imported until they get defined in the Custom Element registry, including all parent classes if any
so if there is index or any other file in between and you omit extension, then it's gonna break the analyzer and it's not gonna find where the class or one of it parent classes are coming from
so as a rule of thumb: use extension, it's not that hard

File: tsconfig.json (hunk 2 above)
Comment (bashmish) [reply tsconfig.json:12]: can you please check if this doesn't break our build of dockit-core
you can use [CLI's `link` command](https://backlight.dev/docs/cli#link) to see the built package

File: box/src/Box.styles.ts [hunk 8]
```diff
@@ -1,9 +1,12 @@
-.wrapper {
+export const wrapper = /*css*/ `
```
Comment (bashmish) [inline box/src/Box.styles.ts:1]: I didn't see this yesterday
why?
you didn't add a Shadow DOM to this component, did you? so this is making things worse from the styles encapsulation perspective

File: box/src/Box.ts [hunk 9]
```diff
@@ -8,12 +8,8 @@ export class Box extends HTMLElement {
...
 
     this.innerHTML = /*html*/ `
-      <div class="${styles.wrapper}">
-        ${
-          hasCheckedBackground
-            ? `<div class="${styles.checkered} ${className}"></div>`
-            : ''
-        }
+      <div style="${style};${styles.wrapper}" class="${className}">
+        ${hasCheckedBackground ? `<div style="${styles.checkered}"></div>` : ''}
```
Comment (bashmish) [inline box/src/Box.ts:12]: `className` was previously on the inner DIV, but you moved it to the parent one
how is this related to TS migration? if you are fixijng some bugs, please create individual PRs with them and clear description why you do it

File: showcases/src/Showcases.ts [hunk 10]
```diff
@@ -39,10 +38,12 @@ export class Showcases extends HTMLElement {
       ${acc}
       ${renderComponent({
         componentClass,
-        [showcaseAttr]: showcase,
         hasLongText,
         captionWidth,
         hasCheckeredBackground,
+        ...(!!showcaseClasses
+          ? { showcaseClass: showcase, showcaseStyle: undefined }
+          : { showcaseClass: undefined, showcaseStyle: showcase }),
```
Comment (bashmish) [inline showcases/src/Showcases.ts:46]: `[showcaseAttr]: showcase,` worked pretty well for me
I checked in VSCode and it seems to be valid TS syntax

File: space/src/Space.ts [hunk 11]
```diff
@@ -1,4 +1,6 @@
 export class Space extends HTMLElement {
+  scale: any[] | Record<any, any>;
```
Comment (bashmish) [inline space/src/Space.ts:2]:
```suggestion
  scale: (number | string)[] | Record<string, number | string>;
```

File: tailwind-showcases/src/TailwindShowcases.ts [hunk 12]
```diff
@@ -73,6 +73,8 @@ const getClassNames = (suffixes) => ({
 });
 
 export class TailwindShowcases extends HTMLElement {
+  theme: Record<any, any>;
```
Comment (bashmish) [inline tailwind-showcases/src/TailwindShowcases.ts:76]: can't you import this type from `tailwindcss`? or make it less wide

File: box/src/Box.styles.ts [hunk 13]
```diff
@@ -18,6 +21,8 @@
   background-position: 0 0, 0 0, -0.5rem -0.5rem, 0.5rem 0.5rem;
   position: absolute;
   top: 0px;
+  width: 100%;
+  height: 100%;
```
Comment (bashmish) [inline box/src/Box.styles.ts:25]: CSS fixes seem unrelated to this PR, normally I expect such fixes to be in individual PRs
here I see a lot of small things like this, this is not good, hard to review, hard to understand from Git history in the future why changes were made

File: showcases/src/Showcases.ts (hunk 10 above)
Comment (bashmish) [inline showcases/src/Showcases.ts:46]: looks like TS has a bug here :) `[showcaseAttr]: showcase,` should just work
but to workaround it there is maybe a better alternative:

```ts
      ${renderComponent({
        componentClass,
        showcaseClass: undefined,
        showcaseStyle: undefined,
        [showcaseAttr]: showcase,
        hasLongText,
        captionWidth,
        hasCheckeredBackground,
      })}
```

Comment (bashmish) [reply showcases/src/Showcases.ts:46]: but to fix this properly it's best to call `renderCaptionedBox` and `renderCaptionedText` separately under condition, instead of assigning method to a var

File: clipboard/src/Clipboard.ts [hunk 14]
```diff
@@ -11,7 +11,7 @@ export class Clipboard extends HTMLElement {
     const text = this.shadowRoot
       .querySelector('slot')
       .assignedNodes()[0]?.textContent;
-    copyTextToClipboard(text);
+    copyTextToClipboard.call(text);
```
Comment (bashmish) [inline clipboard/src/Clipboard.ts:14]: works for me without `.call`, with TS config or without
I think the issue was only with types, not with runtime behavior
it's a function in all cases

**Review by bashmish — CHANGES_REQUESTED** [review]
I think there is a bit too many unrelated fixes in this PR, this makes it hard to review and I can't approve yet. Please create individual PRs for such things, this makes life of everyone easier when reviewing and maintaining code in the long-term.
As for the main change: TS and index files its good, so I could have approved that part already by now.
